README.md
# XOOPS - SCEditor

[![Build Status](https://travis-ci.org/samclarke/SCEditor.png?branch=master)](https://travis-ci.org/samclarke/SCEditor)

## Intro

SCEditor is a WYSIWYG text editor BBCODE which in order to be light and combine all the features required in a professional BBCODE editor

For more information about the editor visit [sceditor.com](http://www.sceditor.com/)

At the moment the Integrated version support XOOPS Smily system and in the future the FileManager.

## Install

simply download the content of the repo in a folder named ```sceditor``` and put it in ```class\xoopseditor\```

Next step is to add some additional BBCodes to XOOPS to have all functions of the editor by replacing the file at ```class\textsanitizer\youtube\youtube.php``` with the one in ```Extra_BBCode```

Now head to the file ```class\module.textsanitizer.php``` and search for these lines

	$patterns[] = "/\[size=(['\"]?)([a-z0-9-]*)\\1](.*)\[\/size\]/sU";
    $replacements[] = '<span style="font-size: \\2;">\\3</span>';

and disable them by adding ```//``` before them, and add these lines after them

	    // sceditor
        $patterns[] = "/\[color=#(['\"]?)([a-zA-Z0-9]*)\\1](.*)\[\/color\]/sU";
        $replacements[] = '<span style="color: #\\2;">\\3</span>';
        $patterns[] = "/\[size=(['\"]?)([a-z-]*)\\1](.*)\[\/size\]/sU";
        $replacements[] = '<span style="font-size: \\2;">\\3</span>';
        $patterns[] = "/\[size=(['\"]?)([0-9]*)\\1](.*)\[\/size\]/sU";
        $replacements[] = '<font size="\\2">\\3</font>';
        $patterns[] = "/\[justify](.*)\[\/justify\]/sU";
        $replacements[] = '<div style="text-align: justify;">\\1</div>';
        $patterns[] = "/\[s](.*)\[\/s\]/sU";
        $replacements[] = '<del>\\1</del>';
        $patterns[] = "/\[sub](.*)\[\/sub\]/sU";
        $replacements[] = '<sub>\\1</sub>';
        $patterns[] = "/\[sup](.*)\[\/sup\]/sU";
        $replacements[] = '<sup>\\1</sup>';
        $patterns[] = "/\[hr]/sU";
        $replacements[] = '<hr/>';
        $patterns[] = "/\[email=(['\"]?)([^'\"<>]*)\\1](.*)\[\/email\]/sU";
        $replacements[] = '<a href="mailto:\\2" title="\\2">\\2</a>';
        $patterns[] = "/\[ol](.*)\[\/ol\]/sU";
        $replacements[] = '<ol>\\1</ol>';
        $patterns[] = "/\[table](.*)\[\/table\]/sU";
        $replacements[] = '<table style="width:auto;margin:0 auto">\\1</table>';
        $patterns[] = "/\[tr](.*)\[\/tr\]/sU";
        $replacements[] = '<tr>\\1</tr>';
        $patterns[] = "/\[td](.*)\[\/td\]/sU";
        $replacements[] = '<td>\\1</td>';

and that's it ;)

## Options

See the documentation of the editor here [options page](http://www.sceditor.com/documentation/options/) for a full list of options.

The lang, rtl and fonts options are changable at the Language files at ```class\xoopseditor\sceditor\language\``` and for a more control over the options at the call head at ```class\xoopseditor\sceditor\formsceditor.php```



## Donate

If you would like to make a donation to the editor project you can via
[PayPal](https://www.paypal.com/cgi-bin/webscr?cmd=_s-xclick&hosted_button_id=AVJSF5NEETYYG)
or via [Flattr](http://flattr.com/thing/400345/SCEditor)




changes.txt
/** Youtube
C:\wamp\www\xo256\class\textsanitizer\youtube\youtube.php
*/

    function load(&$ts)
    {
        $ts->patterns[] = "/\[youtube([^\"]*)\]([^\"]*)\[\/youtube\]/esU";
        $ts->replacements[] = __CLASS__ . "::decode( '\\2' )";
    }

    function decode($url)
    {        
    
        if (preg_match("/^http:\/\/(www\.)?youtube\.com\/watch\?v=(.*)/i", $url, $matches)) {
            $src = "//www.youtube.com/v/" . $matches[2];
        }else{
            $src = "//www.youtube.com/v/" . $url; 
        }

        $code = "<object width='420' height='315'>";
        $code.= "<param name='movie' value='{$src}?version=3'></param>";
        $code.= "<param name='allowFullScreen' value='true'></param>";
        $code.= "<param name='allowscriptaccess' value='always'></param>";
        $code.= "<embed src='{$src}?version=3'";
        $code.= "type='application/x-shockwave-flash'";
        $code.= "width='420'";
        $code.= "height='315'";
        $code.= "allowscriptaccess='always'";
        $code.= "allowfullscreen='true'>";
        $code.= "</embed>";
        $code.= "</object>";
        
        return $code;
    }
    
/** BBCode
C:\wamp\www\xo256\class\module.textsanitizer.php
> comment
$patterns[] = "/\[size=(['\"]?)([a-z0-9-]*)\\1](.*)\[\/size\]/sU";
$replacements[] = '<span style="font-size: \\2;">\\3</span>';
> add
// sceditor
$patterns[] = "/\[color=#(['\"]?)([a-zA-Z0-9]*)\\1](.*)\[\/color\]/sU";
$replacements[] = '<span style="color: #\\2;">\\3</span>';
$patterns[] = "/\[size=(['\"]?)([a-z-]*)\\1](.*)\[\/size\]/sU";
$replacements[] = '<span style="font-size: \\2;">\\3</span>';
$patterns[] = "/\[size=(['\"]?)([0-9]*)\\1](.*)\[\/size\]/sU";
$replacements[] = '<font size="\\2">\\3</font>';
$patterns[] = "/\[justify](.*)\[\/justify\]/sU";
$replacements[] = '<div style="text-align: justify;">\\1</div>';
$patterns[] = "/\[s](.*)\[\/s\]/sU";
$replacements[] = '<del>\\1</del>';
$patterns[] = "/\[sub](.*)\[\/sub\]/sU";
$replacements[] = '<sub>\\1</sub>';
$patterns[] = "/\[sup](.*)\[\/sup\]/sU";
$replacements[] = '<sup>\\1</sup>';

formsceditor.php
<?php
/*
 You may not change or alter any portion of this comment or credits
 of supporting developers from this source code or any supporting source code
 which is considered copyrighted (c) material of the original comment or credit authors.

 This program is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
*/

/**
 *  SCEditor adapter for XOOPS
 *
 * @copyright       The XOOPS Project http://sourceforge.net/projects/xoops/
 * @license         http://www.fsf.org/copyleft/gpl.html GNU public license
 * @package         class
 * @subpackage      editor
 * @since           2.5.6
 * @author          Michael Beck <mambax7@gmail.com>
 * @version         $Id: formtinymce.php 8066 2011-11-06 05:09:33Z beckmi $
 */

xoops_load('XoopsEditor');

class XoopsFormSceditor extends XoopsEditor
{
    var $language = _LANGCODE;
    var $width;
    var $height;
//    var $editor;

    function __construct($configs)
    {
        $current_path = __FILE__;
        if ( DIRECTORY_SEPARATOR != "/" ) {
            $current_path = str_replace( strpos( $current_path, "\\\\", 2 ) ? "\\\\" : DIRECTORY_SEPARATOR, "/", $current_path);
        }
        $this->rootPath = "/class/xoopseditor/sceditor";
        parent::__construct($configs);
        $this->width = $configs['width'];
        $this->height = $configs['height'];
        $this->setLanguage(_XOOPS_EDITOR_SCEDITOR_LANGUAGE);
    }
    function getName()
    {
        return $this->name;
    }
    function setName($value)
    {
        $this->name = $value;
    }
    /**
     * get textarea width
     *
     * @return  string
     */
    function getWidth()
    {
        return $this->width;
    }
    /**
     * get textarea height
     *
     * @return  string
     */
    function getHeight()
    {
        return $this->height;
    }
    /**
     * get language
     *
     * @return  string
     */
    function getLanguage()
    {
        return str_replace('_','-',strtolower($this->language));
    }
    /**
     * set language
     *
     * @return  null
     */
    function setLanguage($lang='en')
    {
        $this->language = $lang;
    }

    /**
     * Renders the Javascript function needed for client-side for validation
     *
     * @return    string
     */
    function renderValidationJS()
    {
        if ($this->isRequired() && $eltname = $this->getName()) {
            $eltcaption = $this->getCaption();
            $eltmsg = empty($eltcaption) ? sprintf( _FORM_ENTER, $eltname ) : sprintf( _FORM_ENTER, $eltcaption );
            $eltmsg = str_replace('"', '\"', stripslashes( $eltmsg ) );
            $ret = "\n";
            $ret.= "if ( tinyMCE.get('{$eltname}').getContent() == \"\" || tinyMCE.get('{$eltname}').getContent() == null)";
            $ret.= "{ window.alert(\"{$eltmsg}\"); myform.{$eltname}.focus(); return false; }";
            return $ret;
            }
        return '';
    }
    /**
     * prepare HTML for output
     *
     * @return  sting HTML
     */
    function render()
    {
      static $isJsLoaded = false;
      $this->rootPath = "/class/xoopseditor/sceditor/sceditor";
      $ret = "\n";
      if(!$isJsLoaded)
      {
        $GLOBALS['xoTheme']->addStylesheet( XOOPS_URL . $this->rootPath . '/minified/themes/default.min.css', array('type'=>'text/css', 'media'=>'all') );
        $GLOBALS['xoTheme']->addScript('browse.php?Frameworks/jquery/jquery.js');
        $GLOBALS['xoTheme']->addScript( XOOPS_URL . $this->rootPath . '/minified/jquery.sceditor.bbcode.min.js' );
        if($this->getLanguage() != 'en')
          $GLOBALS['xoTheme']->addScript( XOOPS_URL . $this->rootPath . '/languages/'.$this->getLanguage().'.js' );
        $isJsLoaded = true;
      }
      
      if(_XOOPS_EDITOR_SCEDITOR_DIRECTION == "ltr"){
        $rtl = "false";
        $toolbar = "bold,italic,underline,strike,subscript,superscript|left,center,right,justify|font,size,color,removeformat|cut,copy,paste,pastetext|bulletlist,orderedlist|table|code,quote|horizontalrule,image,email,link,unlink|emoticon,youtube,date,time|ltr,rtl|print,maximize,source";
      }else{
        $rtl = "true";
        $toolbar = "bold,italic,underline,strike,subscript,superscript|right,center,left,justify|font,size,color,removeformat|cut,copy,paste,pastetext|bulletlist,orderedlist|table|code,quote|horizontalrule,image,email,link,unlink|emoticon,youtube,date,time|rtl,ltr|print,maximize,source";
      }
      
      $smilies_dropdown = '';
      $smilies_more = '';
      foreach(MyTextSanitizer::getInstance()->getSmileys() as $smily){
        if($smily['display'])
          $smilies_dropdown .= '"'.$smily['code'].'": "'.XOOPS_URL.'/uploads/'.$smily['smile_url'].'",';
        else
          $smilies_more .= '"'.$smily['code'].'": "'.XOOPS_URL.'/uploads/'.$smily['smile_url'].'",';
      }
      $ret.= "<script type='text/javascript'>\n";
      $ret.= "jQuery(document).ready(function(){\n";
			$ret.= "  $('#".$this->getName()."').sceditor({\n";
      $ret.= "    rtl : ".$rtl.",\n";
      $ret.= "    fonts : \""._XOOPS_EDITOR_SCEDITOR_FONTS."\",\n";
      $ret.= "    toolbar : \"".$toolbar."\",\n";
      $ret.= "    locale : '".$this->getLanguage()."',\n";
			$ret.= "    plugins: 'bbcode',\n";
			$ret.= "    style: '".XOOPS_URL . $this->rootPath . "/minified/jquery.sceditor.default.min.css',\n";
			$ret.= "    emoticons: {dropdown: {".rtrim($smilies_dropdown, ",")."},more: {".rtrim($smilies_more, ",")."}}\n";
			$ret.= "  });\n";
      $ret.= "});\n";
      $ret.= "</script>\n";

      $ret.= "<textarea class='".$this->getName()."' name='".$this->getName()."' id='".$this->getName()."' ".$this->getExtra()."style='width:".$this->getWidth().";height:".$this->getHeight().";'>" . $this->getValue() . "</textarea>";
         return $ret ;

    }
}
?>

language/arabic.php
<?php
/**
 *  TinyMCE adapter for XOOPS
 *
 * @copyright   The XOOPS project http://www.xoops.org/
 * @license     http://www.fsf.org/copyleft/gpl.html GNU public license
 * @package		core
 * @subpackage	xoopseditor
 * @since       2.3.0
 * @author		Taiwen Jiang <phppp@users.sourceforge.net>
 * @version		$Id: english.php 10054 2012-08-11 10:06:44Z beckmi $
 */
/*
 * Assocated with editor_registry.php
 */
// Name of the editor
define("_XOOPS_EDITOR_SCEDITOR","SCEditor");

// The file must code.js must be in the class\xoopseditor\sceditor\sceditor\languages
define("_XOOPS_EDITOR_SCEDITOR_LANGUAGE", "ar");

// DIRECTION , ltr or rtl
define("_XOOPS_EDITOR_SCEDITOR_DIRECTION","rtl");

// FONT LIST, FORMAT: "Name=value1,value2;Name=value"
define("_XOOPS_EDITOR_SCEDITOR_FONTS","Arial,Arial Black,Comic Sans MS,Courier New,Georgia,Impact,Sans-serif,Serif,Times New Roman,Tahoma,Trebuchet MS,Verdana");
?>

language/english.php
<?php
/**
 *  TinyMCE adapter for XOOPS
 *
 * @copyright   The XOOPS project http://www.xoops.org/
 * @license     http://www.fsf.org/copyleft/gpl.html GNU public license
 * @package		core
 * @subpackage	xoopseditor
 * @since       2.3.0
 * @author		Taiwen Jiang <phppp@users.sourceforge.net>
 * @version		$Id: english.php 10054 2012-08-11 10:06:44Z beckmi $
 */
/*
 * Assocated with editor_registry.php
 */
// Name of the editor
define("_XOOPS_EDITOR_SCEDITOR","SCEditor");

// The file must code.js must be in the class\xoopseditor\sceditor\sceditor\languages
define("_XOOPS_EDITOR_SCEDITOR_LANGUAGE","en");

// DIRECTION , ltr or rtl
define("_XOOPS_EDITOR_SCEDITOR_DIRECTION","ltr");

// FONT LIST, FORMAT: "Name=value1,value2;Name=value"
define("_XOOPS_EDITOR_SCEDITOR_FONTS","Arial,Arial Black,Comic Sans MS,Courier New,Georgia,Impact,Sans-serif,Serif,Times New Roman,Trebuchet MS,Verdana");
?>

sceditor/languages/ar.js
/**
 * @author Atramez_Zeton http://onyx-sy.net
 * @license [MIT](http://www.opensource.org/licenses/mit-license.php)
 */
(function ($) {
	'use strict';

	$.sceditor.locale["ar"] = {
		"Bold": "عريض",
		"Italic": "مائل",
		"Underline": "خط من الأسفل",
		"Strikethrough": "خط في المنتصف",
		"Subscript": "حرف منخفض",
		"Superscript": "حرف مرتفع",
		"Align left": "انحياز إلى اليسار",
		"Center": "توسط",
		"Align right": "انحياز إالى اليمين",
		"Justify": "ملأ السطر",
		"Font Name": "نوع الخط",
		"Font Size": "حجم الخط",
		"Font Color": "لون الخط",
		"Remove Formatting": "ازالة التعديلات",
		"Cut": "قص",
		"Your browser does not allow the cut command. Please use the keyboard shortcut Ctrl/Cmd-X": "Ctrl/Cmd-X متصفحك لا يدعم اوامر القص الرجاء استخدام اختصارات لوحة التحكم",
		"Copy": "نسخ",
		"Your browser does not allow the copy command. Please use the keyboard shortcut Ctrl/Cmd-C": "Ctrl/Cmd-C متصفحك لا يدعم اوامر النسخ الرجاء استخدام اختصارات لوحة التحكم",
		"Paste": "لصق",
		"Your browser does not allow the paste command. Please use the keyboard shortcut Ctrl/Cmd-V": "Ctrl/Cmd-V متصفحك لا يدعم اوامر اللصق الرجاء استخدام اختصارات لوحة التحكم",
		"Paste your text inside the following box:": "قم بلصق نصّك في المربع",
		"Paste Text": "الصق النص",
		"Bullet list": "قائمة نقطية",
		"Numbered list": "قائمة مرقمة",
		"Undo": "تراجع",
		"Redo": "تقدم",
		"Rows:": "اسطر",
		"Cols:": "اعمدة",
		"Insert a table": "ادرج جدول",
		"Insert a horizontal rule": "ادرج مسطرة افقية",
		"Code": "كود",
		"Width (optional):": "عرض (اختياري)",
		"Height (optional):": "ارتفاع (اختياري)",
		"Insert an image": "ادرج صورة",
		"E-mail:": "بريد الكتروني",
		"Insert an email": "ادرج بريدا الكترونيا",
		"URL:": "وصلة موقع",
		"Insert a link": "ادرج وصلة لموقع",
		"Unlink": "ازالة الوصلة",
		"More": "المزيد",
		"Insert an emoticon": "ادرج وجها",
		"Video URL:": "وصلة فيديو",
		"Insert": "ادرج",
		"Insert a YouTube video": "ادرج وصلة فيديو يوتيوب",
		"Insert current date": "ادرج التاريخ الحالي",
		"Insert current time": "ادرج الوقت الحالي",
		"Print": "اطبع",
		"View source": "اظهر المصدر",
		"Description (optional):": "الوصف (اختياري)",
		"Enter the image URL:": "ضع وصلة الصورة",
		"Enter the e-mail address:": "ضع عنوان البريد الإلكتروني",
		"Enter the displayed text:": "ضع النص الذي تريد اظهاره",
		"Enter URL:": "ضع وصلة موقع",
		"Enter the YouTube video URL or ID:": "ضع وصلة فيديو يوتيوب او رقم الفيديو",
		"Insert a Quote":"ادرج اقتباسا",
		"Invalid YouTube video": "هذا الفيديو غير صالح",
		"Left-to-Right": "يسار إلى يمين",
		"Right-to-Left": "يمين إلى يسار",
		"Maximize": "تكبير",

		dateFormat: "day-month-year"
	};
})(jQuery);


Extra_BBCode/youtube.php
<?php
/**
 * TextSanitizer extension
 *
 * You may not change or alter any portion of this comment or credits
 * of supporting developers from this source code or any supporting source code
 * which is considered copyrighted (c) material of the original comment or credit authors.
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 *
 * @copyright       The XOOPS Project http://sourceforge.net/projects/xoops/
 * @license         GNU GPL 2 (http://www.gnu.org/licenses/old-licenses/gpl-2.0.html)
 * @package         class
 * @subpackage      textsanitizer
 * @since           2.3.0
 * @author          Taiwen Jiang <phppp@users.sourceforge.net>
 * @version         $Id: youtube.php 8066 2011-11-06 05:09:33Z beckmi $
 */

class MytsYoutube extends MyTextSanitizerExtension
{
    function encode($textarea_id)
    {
        $config = parent::loadConfig( dirname(__FILE__) );
        $code = "<img src='{$this->image_path}/youtube.gif' alt='" . _XOOPS_FORM_ALTYOUTUBE . "' onclick='xoopsCodeYoutube(\"{$textarea_id}\",\"" . htmlspecialchars(_XOOPS_FORM_ENTERYOUTUBEURL, ENT_QUOTES) . "\",\"" . htmlspecialchars(_XOOPS_FORM_ALT_ENTERHEIGHT, ENT_QUOTES) . "\",\"" . htmlspecialchars(_XOOPS_FORM_ALT_ENTERWIDTH, ENT_QUOTES) . "\");'  onmouseover='style.cursor=\"hand\"'/>&nbsp;";
        $javascript = <<<EOH
            function xoopsCodeYoutube(id, enterFlashPhrase, enterFlashHeightPhrase, enterFlashWidthPhrase)
            {
                var selection = xoopsGetSelect(id);
                if (selection.length > 0) {
                    var text = selection;
                } else {
                    var text = prompt(enterFlashPhrase, "");
                }
                var domobj = xoopsGetElementById(id);
                if ( text.length > 0 ) {
                    var text2 = prompt(enterFlashWidthPhrase, "425");
                    var text3 = prompt(enterFlashHeightPhrase, "350");
                    var result = "[youtube="+text2+","+text3+"]" + text + "[/youtube]";
                    xoopsInsertText(domobj, result);
                }
                domobj.focus();
            }
EOH;

        return array($code, $javascript);
    }

    function load(&$ts)
    {
        $ts->patterns[] = "/\[youtube([^\"]*)\]([^\"]*)\[\/youtube\]/esU";
        $ts->replacements[] = __CLASS__ . "::decode( '\\2' )";
    }

    function decode($url)
    {        
    
        if (preg_match("/^http:\/\/(www\.)?youtube\.com\/watch\?v=(.*)/i", $url, $matches)) {
            $src = "//www.youtube.com/v/" . $matches[2];
        }else{
            $src = "//www.youtube.com/v/" . $url; 
        }

        $code = "<object width='420' height='315'>";
        $code.= "<param name='movie' value='{$src}?version=3'></param>";
        $code.= "<param name='allowFullScreen' value='true'></param>";
        $code.= "<param name='allowscriptaccess' value='always'></param>";
        $code.= "<embed src='{$src}?version=3'";
        $code.= "type='application/x-shockwave-flash'";
        $code.= "width='420'";
        $code.= "height='315'";
        $code.= "allowscriptaccess='always'";
        $code.= "allowfullscreen='true'>";
        $code.= "</embed>";
        $code.= "</object>";
        
        return $code;
    }
}
?>